ubuntu_image/parser.py

```diff
@@ -320,41 +332,56 @@ def parse(stream_or_string):
             offset_write = structure.get('offset-write')
             size = structure['size']
             structure_type = structure['type']
-            # Validate structure types.  These can be either GUIDs, two hex
-            # digits, hybrids, or the special 'mbr' type.  The basic syntactic
-            # validation happens above in the Voluptuous schema, but here we
-            # need to ensure cross-attribute constraints.  Specifically,
-            # hybrids and 'mbr' are allowed for either schema, but GUID-only
-            # is only allowed for GPT, while 2-digit-only is only allowed for
-            # MBR.  Note too that 2-item tuples are also already ensured.
+            structure_role = structure.get('role')
+            # Validate structure types and roles. These can be either GUIDs,
+            # two hex digits or hybrids.  The basic syntactic validation
+            # happens above in the Voluptuous schema, but here we need to
+            # ensure cross-attribute constraints.  Specifically, hybrids are
+            # allowed for either schema, but GUID-only is only allowed for GPT,
+            # while 2-digit-only is only allowed for MBR roles.
+            # Note too that 2-item tuples are also already ensured.
+            # We also still support the deprecated special type 'mbr' for
+            # legacy purposes, but issue a warning
+            if structure_type == 'mbr':
+                warn("volumes:<volume name>:structure:<N>:type = 'mbr' is "
+                     'deprecated; use role instead', DeprecationWarning)
+                if structure_role:
+                    raise GadgetSpecificationError(
+                        'Type mbr and role fields assigned at the same time, '
+                        'please use the mbr role instead')
+                structure_role = StructureRole.mbr
             if (isinstance(structure_type, UUID) and
                     schema is not VolumeSchema.gpt):
                 raise GadgetSpecificationError(
                     'MBR structure type with non-MBR schema')
             elif (isinstance(structure_type, str) and
```

Comment:
It will be nice when we can get rid of `type='mbr'` so that the logic here can be simplified.  I had to work out the cases in my head again to convince myself that these checks are still correct, and I think they are.  I'd probably add a comment above this to explain (or remind us ;) what's going on, but don't worry about that.  I will do this after the branch lands.